fakestorage/upload.go

```diff
@@ -136,72 +152,40 @@ func (s *Server) insertFormObject(r *http.Request) xmlResponse {
 		},
 		Content: infile,
 	}
-	obj, err = s.createObject(obj)
+	obj, err = s.createObject(obj, backend.NoConditions{})
 	if err != nil {
 		return xmlResponse{errorMessage: err.Error()}
 	}
 	defer obj.Close()
 	return xmlResponse{status: http.StatusNoContent}
 }
 
-func (s *Server) checkUploadPreconditions(r *http.Request, bucketName string, objectName string) *jsonResponse {
+func (s *Server) wrapUploadPreconditions(r *http.Request, bucketName string, objectName string) (*generationCondition, error) {
+	result := &generationCondition{
```

Comment:
I think we can stop using a pointer here to avoid the dereference later. This is guaranteed to basically never be nil, so no need to use a pointer.
```suggestion
func (s *Server) wrapUploadPreconditions(r *http.Request, bucketName string, objectName string) (generationCondition, error) {
	result := generationCondition{
```

Then we can always return result when there's an error too.